src/passes/GlobalStructInference.cpp

```diff
@@ -211,16 +214,98 @@ struct GlobalStructInference : public Pass {
       std::sort(globals.begin(), globals.end());
     }
 
-    // Optimize based on the above.
-    struct FunctionOptimizer
-      : public WalkerPass<PostWalker<FunctionOptimizer>> {
-      bool isFunctionParallel() override { return true; }
+    // We are looking for the case where we can pick between two values using a
+    // single comparison. More than two values, or more than a single
+    // comparison, lead to tradeoffs that may not be worth it.
+    //
+    // Note that situation may involve more than two globals. For example we may
+    // have three relevant globals, but two may have the same value. In that
+    // case we can compare against the third:
+    //
+    //  $global0: (struct.new $Type (i32.const 42))
+    //  $global1: (struct.new $Type (i32.const 42))
+    //  $global2: (struct.new $Type (i32.const 1337))
+    //
+    // (struct.get $Type (ref))
+    //   =>
+    // (select
+    //   (i32.const 1337)
+    //   (i32.const 42)
+    //   (ref.eq (ref) $global2))
+    //
+    // To discover these situations, we compute and group the possible values
+    // that can be read from a particular struct.get, using the following data
+    // structure.
+    struct Value {
+      // A value is either a constant, or if not, then we point to whatever
+      // expression it is.
+      std::variant<PossibleConstantValues, Expression*> content;
+      // The list of globals that have this Value. In the example from above,
+      // the Value for 42 would list globals = [$global0, $global1].
+      // TODO: SmallVector?
+      std::vector<Name> globals;
+
+      bool isConstant() const {
+        return std::get_if<PossibleConstantValues>(&content);
+      }
 
-      std::unique_ptr<Pass> create() override {
-        return std::make_unique<FunctionOptimizer>(parent);
+      const PossibleConstantValues& getConstant() const {
```

Comment:
Hmm, yeah, that might be a little less code here. But I'm not sure if it would be more clear (as working with pointers is less nice). Also, this is parallel to other existing APIs:

https://github.com/WebAssembly/binaryen/blob/1079a9e34599e65ee25fb5f32caa57bd21737593/src/ir/possible-constant.h#L138-L154

Reply:
I see we have divergent styles here. All of the variant-using code I've written uses the pointer approach :)